drivers: sensor: Add tmd2620

pefech opened the pull request:
This PR adds the driver for the IR proximity sensor tmd2620 made by ams-osram, including a sample.
The driver works in polling and trigger mode with power management capabilities and its possible to configure almost every setting of the sensor via Kconfig / Menuconfig. Every working mode can be tested with the sample.

You can find datasheets of the sensor here: [TMD2620 datasheet](https://ams.com/documents/20143/36005/TMD2620_DS000461_3-00.pdf)

MaureenHelm commented on drivers/sensor/tmd2620/Kconfig:
```diff
@@ -0,0 +1,122 @@
+	int "Proximity Interrupt Low Threshold"
+	default 0
+	range 0 255
+
+config TMD2620_PIHT
+	int "Proximity Interrupt High Threshold"
+	default 1 if !TMD2620_TRIGGER_GLOBAL_THREAD
+	default 127 if TMD2620_TRIGGER_GLOBAL_THREAD
+	range 0 255
+
+config TMD2620_PPERS
+	int "Proximity Interrupt Filter"
+	help
+		0: every proximity cycle generates an interrupt
+		1: 1 consecutive proximity value out of threshold range
+		2: 2 consecutive proximity values out of threshold range
+		...
+	default 0 if !TMD2620_TRIGGER_GLOBAL_THREAD
+	default 7 if TMD2620_TRIGGER_GLOBAL_THREAD
+	range 0 15
+
+config TMD2620_WAIT_ENABLE
+	bool "Enable wait mode"
+	default n
+
+config TMD2620_WTIME
+	int "Wait Time"
+	help
+		0: disable wait timer
+		1: 2.78ms
+		2: 5.56ms
+		...
+	default 255
+	range 1 256 # decrement by 1
+
+config TMD2620_WLONG
+	bool "Wait Long"
+	help
+		increase wait time by a factor of 12x
+	default n
```
Please convert these to devicetree properties to allow per-instance configuration

MaureenHelm commented on drivers/sensor/tmd2620/tmd2620.c:
```diff
@@ -0,0 +1,441 @@
+
+PM_DEVICE_DT_INST_DEFINE(0, tmd2620_pm_action);
+
+DEVICE_DT_INST_DEFINE(0, tmd2620_init, PM_DEVICE_DT_INST_GET(0),
+			&tmd2620_data, &tmd2620_config, POST_KERNEL,
+			CONFIG_SENSOR_INIT_PRIORITY, &tmd2620_driver_api);
```
Please use `DT_INST_FOREACH_STATUS_OKAY` to support multiple instances

MaureenHelm reviewed (changes requested):
Thank you for the contribution.

Please add this sensor to `tests/drivers/build_all/sensor/` and modify the sample to be independent of the specific driver (i.e., a generic proximity sensor sample).

There also seems to be an overabundance of tabs to indent lines. You should configure your text editor indent size to 8.

pefech commented:
Hi, thank you for reviewing so quickly!
I made the driver support multiple instances and changed the configuration for the sensor settings to be configured in the device tree.
But I'm struggling a little bit with writing a generic proximity sensor sample, because the most drivers use Kconfig to be configured in trigger or polling mode (like the tmd2620 driver too). I'm not sure how to implement a statement into the sample that is able to detect if the driver has to be used in trigger or polling mode because the Kconfig entries are all diffrent. For example the apds9960 uses ```CONFIG_APDS9960_TRIGGER``` and my implementation for the tmd2620 uses ```CONFIG_TMD2620_TRIGGER```.

The only option that comes to my mind would be to have two samples - one for polling and one for trigger mode.
Or do I miss something important?

pefech committed:
drivers: sensor: Add tmd2620 driver

Adds tmd2620 driver and devicetree bindings to work in
trigger and polling mode supporting Power management.

Signed-off-by: Peter Fecher <p.fecher@phytec.de>

pefech committed:
samples: sensor: Add proximity sensor sample

Adds generic sample for proximity sensors.

Signed-off-by: Peter Fecher <p.fecher@phytec.de>

pefech commented:
The last push includes all your requested changes.

gmarull commented on samples/sensor/proximity_polling/src/main.c:
```diff
@@ -0,0 +1,61 @@
+		k_sleep(K_MSEC(2000));
+		print_prox_data();
+
+#ifdef CONFIG_PM_DEVICE
+		pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+		printk("set low power state for 2s\n");
+		k_sleep(K_MSEC(2000));
+		pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);
+#endif
```
please don't use this API in applications, you should be using runtime PM framework instead.

pefech commented on samples/sensor/proximity_polling/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+      regex:
+        - "Proximity sensor sample application"
+        - "Found [0-9] proximity sensor(s): (.*)$"
+        - "Proximity on ([^:]*): [0-9]*$"
```
If I understood it correctly the proximity value is an adimensional value from 1 to 255 where 1 is near and 255 is far.
It has been described here like this: [sensor_channel](https://docs.zephyrproject.org/apidoc/latest/group__sensor__interface.html#gaaa1b502bc029b10d7b23b0a25ef4e934)

It also depends on the settings of the sensor itself (and maybe the material it is supposed to detect) how far an objects proximity would be converted to a unit.

pefech commented on dts/bindings/sensor/ams,tmd2620.yaml:
```diff
@@ -0,0 +1,86 @@
+      type: boolean
+      description: Enables wait mode
+
+    wait-time:
```
the units is specified in the description of ```wait-time```. I'll rename it to ```wait-time-factor```.
